### PR TITLE
oxidize numeric.poly_*

### PR DESCRIPTION
**Pull request by joostvanzwieten**
This PR translates the `numeric.poly_*` functions to Rust and introduces a compact representation for polynomials.

### CODE REVIEW

File: nutils/element.py [hunk 1]
```diff
@@ -440,27 +441,23 @@ def _integer_barycentric_coordinates(self, degree):
...
+        coeffs = numpy.zeros((ndofs, poly.ncoeffs(self.ndims, degree)), dtype=float)
+        powers = self._integer_barycentric_coordinates(degree)
+        for i, p in enumerate(powers):
+            for j, q in enumerate(powers[::-1]):
+                q_sub_p = tuple(map(operator.sub, q, p))
+                if all(power >= 0 for power in q_sub_p[1:]):
+                    coeffs[i, j] = (-1)**q_sub_p[0]*math.factorial(degree)//util.product(map(math.factorial, (q[0], *p[1:], *q_sub_p[1:])))
         assert i == ndofs - 1
-        return types.frozenarray(coeffs, copy=False)
+        return types.frozenarray(coeffs, copy=True)
```
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:452]: Changed to false.

File: nutils/element.py [hunk 2]
```diff
@@ -736,7 +733,11 @@ def get_ndofs(self, degree):
         return self.ref1.get_ndofs(degree)*self.ref2.get_ndofs(degree)
 
     def get_poly_coeffs(self, basis, **kwargs):
-        return numeric.poly_outer_product(self.ref1.get_poly_coeffs(basis, **kwargs), self.ref2.get_poly_coeffs(basis, **kwargs))
+        coeffs1 = self.ref1.get_poly_coeffs(basis, **kwargs)
+        coeffs2 = self.ref2.get_poly_coeffs(basis, **kwargs)
+        coeffs = poly.mul(coeffs1[:,None], coeffs2[None], (poly.MulVar.Left,) * self.ref1.ndims + (poly.MulVar.Right,) * self.ref2.ndims)
+        coeffs = coeffs.reshape(-1, coeffs.shape[-1])
+        return types.frozenarray(coeffs, dtype=float, copy=True)
```
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:740]: Yes!

File: nutils/element.py [hunk 3]
```diff
@@ -736,7 +733,11 @@ def get_ndofs(self, degree):
         return self.ref1.get_ndofs(degree)*self.ref2.get_ndofs(degree)
 
     def get_poly_coeffs(self, basis, **kwargs):
-        return numeric.poly_outer_product(self.ref1.get_poly_coeffs(basis, **kwargs), self.ref2.get_poly_coeffs(basis, **kwargs))
+        coeffs1 = self.ref1.get_poly_coeffs(basis, **kwargs)
+        coeffs2 = self.ref2.get_poly_coeffs(basis, **kwargs)
+        coeffs = poly.mul(coeffs1[:,None], coeffs2[None], (poly.MulVar.Left,) * self.ref1.ndims + (poly.MulVar.Right,) * self.ref2.ndims)
```
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:738]: Done.

File: nutils/element.py [hunk 4]
```diff
@@ -440,27 +441,23 @@ def _integer_barycentric_coordinates(self, degree):
...
-            for q in itertools.product(*[range(degree+1)]*self.ndims):
-                if sum(p+q) <= degree:
-                    coeffs[(i,)+tuple(map(operator.add, p, q))] = (-1)**sum(q)*math.factorial(degree)//util.product(map(math.factorial, (degree-sum(p+q), *p, *q)))
+        coeffs = numpy.zeros((ndofs, poly.ncoeffs(self.ndims, degree)), dtype=float)
+        powers = self._integer_barycentric_coordinates(degree)
+        for i, p in enumerate(powers):
+            for j, q in enumerate(powers[::-1]):
+                q_sub_p = tuple(map(operator.sub, q, p))
+                if all(power >= 0 for power in q_sub_p[1:]):
+                    coeffs[i, j] = (-1)**q_sub_p[0]*math.factorial(degree)//util.product(map(math.factorial, (q[0], *p[1:], *q_sub_p[1:])))
```
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:450]: Yes, but maybe after we have a better picture of what kind of polynomials we want to work with (power series as we are using now, or Bernstein or Legendre).

File: nutils/element.py [hunk 5]
```diff
@@ -440,27 +441,23 @@ def _integer_barycentric_coordinates(self, degree):
...
+            return types.frozenarray(numpy.ones((1, 1), dtype=float), copy=True)
         else:
             P = numpy.array(tuple(self._integer_barycentric_coordinates(degree)), dtype=int)[:, 1:]
-            coeffs_ = numpy.linalg.inv(((P[:, _, :]/degree)**P[_, :, :]).prod(-1))
-            coeffs = numpy.zeros((len(P), *[degree+1]*self.ndims), dtype=float)
-            for i, p in enumerate(P):
-                coeffs[(slice(None), *p)] = coeffs_[i]
-        return types.frozenarray(coeffs, copy=False)
+            coeffs = numpy.linalg.inv(((P[_, :, :]/degree)**P[::-1, _, :]).prod(-1))
+            return types.frozenarray(coeffs, copy=True)
```
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:460]: See the above conversation.

**Review by gertjanvanzwieten — APPROVED** [review]

File: nutils/evaluable.py [hunk 6]
```diff
@@ -3618,87 +3619,270 @@ def _intbounds_impl(self):
...
+        assert isinstance(coeffs, Array) and coeffs.dtype == float and coeffs.ndim >= 1, f'coeffs={coeffs!r}'
+        assert isinstance(points, Array) and points.dtype == float and points.ndim >= 1 and points.shape[-1].isconstant, f'points={points!r}'
         self.points_ndim = int(points.shape[-1])
-        ndim = coeffs.ndim - self.points_ndim
-        assert ndim >= 0, 'argument `coeffs` should have at least one axis per spatial dimension'
         self.coeffs = coeffs
         self.points = points
-        self.ngrad = ngrad
-        super().__init__(args=(points, coeffs), shape=points.shape[:-1]+coeffs.shape[:ndim]+(constant(self.points_ndim),)*ngrad, dtype=float)
+        degree = PolyDegree(coeffs.shape[-1], self.points_ndim)
```
Comment (gertjanvanzwieten) [inline nutils/evaluable.py:3653]: looks like this line can be removed

**Review by gertjanvanzwieten — APPROVED** [review]

File: nutils/evaluable.py (hunk 6 above)
Comment (joostvanzwieten) [reply nutils/evaluable.py]: removed

**Review by gertjanvanzwieten — APPROVED** [review]

**Pull request merged**